velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
   return result.first->task();
 }
 
-// Special function to compare multisets with vectors of variants in a way that
-// we compare all floating point values inside using 'epsilon' constant.
-// Returns true if equal.
+// Tranform a MaterializedRow vector by moving floating-point columns to the end
+// of every row. The number of columns per row and the position of
+// floating-point columns are given by width and floatColumnIndices.
+void moveFloatColumnsToEnd(
+    const std::vector<velox::column_index_t>& floatColumnIndices,
+    size_t width,
+    std::vector<MaterializedRow>& data) {
+  for (auto i = 0; i < data.size(); ++i) {
+    velox::column_index_t lastNonFloatColumn = width - 1;
+    for (auto j = (int32_t)floatColumnIndices.size() - 1; j >= 0; --j) {
+      if (floatColumnIndices[j] != lastNonFloatColumn) {
+        std::swap(data[i][floatColumnIndices[j]], data[i][lastNonFloatColumn]);
+      }
+      --lastNonFloatColumn;
+    }
+  }
+}
+
+// Return true if currentRow and previsouRow are the same at the first width
+// columns.
+bool equalPrefix(
+    const MaterializedRow& currentRow,
+    const MaterializedRow& previousRow,
+    size_t width) {
+  for (auto j = 0; j < width; ++j) {
+    if (currentRow[j] != previousRow[j]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Return true if there are multiple rows with the same prefix of the first
+// width columns. The argument groupedBuckets should be sorted by the first
+// width columns.
+bool hasMultiRowGroups(
+    const std::vector<MaterializedRow>& groupedBuckets,
+    size_t width) {
+  for (auto i = 1; i < groupedBuckets.size(); ++i) {
+    if (equalPrefix(groupedBuckets[i], groupedBuckets[i - 1], width)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Try to bucket rows in left and right into leftBucketed and rightBucketed
+// and returns whether the bucketization is successful. The bucketization is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
+//   same types.
+//   2. Rows in left and right are grouped by non-floating-point column. If
+//   there are multiple floating-point columns, either left or right must have a
+//   single row in every group. This is not required in case of zero or one
+//   floating-point column.
+// The bucket results are stored in leftBucketed and rightBucketed with the
+// floating-point columns moved to the end and rows in ascending order.
+bool tryBucketMaterializedRows(
+    const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right,
+    std::vector<MaterializedRow>& leftBucketed,
+    std::vector<MaterializedRow>& rightBucketed) {
+  if (left.empty() || right.empty() ||
+      left.begin()->size() != right.begin()->size()) {
+    return false;
+  }
+
+  std::vector<velox::column_index_t> floatColumnIndices;
```

Comment:
I still feel that it would be more readable to extract the logic of computing the list of floating-point column indices into a separate function.